File: sdk/python/src/rocksky/filter.py
"""Fluent builder for RSQL filter expressions.

Accepted by the ``filter`` parameter of the catalog and scrobble-feed queries
(``app.rocksky.song.getSongs``, ``app.rocksky.artist.getArtists``,
``app.rocksky.album.getAlbums``, ``app.rocksky.scrobble.getScrobbles``)::

    from rocksky import AppView, Filter

    flt = (
        Filter.eq("artist", "Daft Punk")
        .and_(Filter.gt("duration", 200_000))
        .or_(Filter.is_in("genre", ["house", "electro"]))
    )
    songs = AppView().catalog_songs(50, 0, None, flt.build())
    # artist=="Daft Punk";duration=gt=200000,genre=in=(house,electro)

String values are quoted and escaped automatically when they contain characters
RSQL reserves; ``*`` wildcards pass through unquoted so
``Filter.eq("artist", "Daft*")`` performs a case-insensitive match.

Filterable fields per endpoint:

- songs: ``title, artist, album, albumArtist, genre, composer, label, duration,
  trackNumber, discNumber, mbId, isrc, sha256, uri, albumUri, artistUri, createdAt``
- albums: ``title, artist, year, releaseDate, sha256, uri, artistUri, createdAt``
- artists: ``name, genres, bornIn, born, died, sha256, uri, createdAt``
- scrobbles: ``uri, date, timestamp, title, artist, album, track.title,
  track.artist, track.album, track.albumArtist, track.genre, track.duration,
  track.isrc, track.mbId, user.did, user.handle, user.displayName, artist.name,
  artist.genres``
"""

from __future__ import annotations

import re
from collections.abc import Iterable

FilterValue = str | int | float | bool

_SAFE_VALUE = re.compile(r"^[A-Za-z0-9_.:@*+-]+$")

_COMPARISON = "comparison"
_AND = "and"
_OR = "or"


def _render(value: FilterValue) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return str(value)
    if value and _SAFE_VALUE.match(value):
        return value
    escaped = value.replace("\\", "\\\\").replace('"', '\\"')
    return f'"{escaped}"'


class Filter:
    """An immutable RSQL filter expression. See the module docs for the grammar."""

    __slots__ = ("_expr", "_kind")

    def __init__(self, expr: str, kind: str) -> None:
        self._expr = expr
        self._kind = kind

    # -- comparisons -------------------------------------------------------

    @staticmethod
    def eq(field: str, value: FilterValue) -> Filter:
        """``field==value`` — equals; ``*`` in string values is a wildcard."""
        return Filter(f"{field}=={_render(value)}", _COMPARISON)

    @staticmethod
    def ne(field: str, value: FilterValue) -> Filter:
        """``field!=value`` — not equals."""
        return Filter(f"{field}!={_render(value)}", _COMPARISON)

    @staticmethod
    def gt(field: str, value: FilterValue) -> Filter:
        """``field=gt=value`` — greater than."""
        return Filter(f"{field}=gt={_render(value)}", _COMPARISON)

    @staticmethod
    def ge(field: str, value: FilterValue) -> Filter:
        """``field=ge=value`` — greater than or equal."""
        return Filter(f"{field}=ge={_render(value)}", _COMPARISON)

    @staticmethod
    def lt(field: str, value: FilterValue) -> Filter:
        """``field=lt=value`` — less than."""
        return Filter(f"{field}=lt={_render(value)}", _COMPARISON)

    @staticmethod
    def le(field: str, value: FilterValue) -> Filter:
        """``field=le=value`` — less than or equal."""
        return Filter(f"{field}=le={_render(value)}", _COMPARISON)

    @staticmethod
    def _list(field: str, op: str, values: Iterable[FilterValue]) -> Filter:
        rendered = [_render(v) for v in values]
        if not rendered:
            raise ValueError(f"Filter {op} on {field!r} needs at least one value")
        return Filter(f"{field}{op}({','.join(rendered)})", _COMPARISON)

    @staticmethod
    def is_in(field: str, values: Iterable[FilterValue]) -> Filter:
        """``field=in=(a,b)`` — matches any of the values."""
        return Filter._list(field, "=in=", values)

    @staticmethod
    def is_out(field: str, values: Iterable[FilterValue]) -> Filter:
        """``field=out=(a,b)`` — matches none of the values."""
        return Filter._list(field, "=out=", values)

    @staticmethod
    def is_null(field: str) -> Filter:
        """``field==null`` — the field is NULL."""
        return Filter(f"{field}==null", _COMPARISON)

    @staticmethod
    def is_not_null(field: str) -> Filter:
        """``field!=null`` — the field is not NULL."""
        return Filter(f"{field}!=null", _COMPARISON)

    # -- combinators -------------------------------------------------------

    def _wrapped(self) -> str:
        return f"({self._expr})" if self._kind == _OR else self._expr

    def and_(self, other: Filter) -> Filter:
        """Both sides must match (``;``). An ``or`` operand is parenthesized."""
        return Filter(f"{self._wrapped()};{other._wrapped()}", _AND)

    def or_(self, other: Filter) -> Filter:
        """Either side may match (``,``)."""
        return Filter(f"{self._expr},{other._expr}", _OR)

    # -- output ------------------------------------------------------------

    def build(self) -> str:
        """The RSQL expression string to send as the ``filter`` query param."""
        return self._expr

    def __str__(self) -> str:
        return self._expr

    def __repr__(self) -> str:
        return f"Filter({self._expr!r})"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Filter):
            return NotImplemented
        return self._expr == other._expr and self._kind == other._kind

    def __hash__(self) -> int:
        return hash((self._expr, self._kind))

File: sdk/go/rocksky/filter.go
package rocksky

// Fluent builder for RSQL filter expressions, accepted by the `filter`
// parameter of the catalog and scrobble-feed queries
// (app.rocksky.song.getSongs, app.rocksky.artist.getArtists,
// app.rocksky.album.getAlbums, app.rocksky.scrobble.getScrobbles).
//
//	f := rocksky.Eq("artist", "Daft Punk").
//		And(rocksky.Gt("duration", 200000)).
//		Or(rocksky.In("genre", "house", "electro"))
//	songs, _ := client.CatalogSongs(ctx, 50, 0, "", f.Build())
//	// artist=="Daft Punk";duration=gt=200000,genre=in=(house,electro)
//
// String values are quoted and escaped automatically when they contain
// characters RSQL reserves; `*` wildcards pass through unquoted so
// Eq("artist", "Daft*") performs a case-insensitive match.

import (
	"fmt"
	"strconv"
	"strings"
)

// Expression node kinds. AND-combining parenthesizes an operand iff it is an
// OR node, preserving RSQL precedence (';' binds tighter than ',').
type filterKind int

const (
	kindComparison filterKind = iota
	kindAnd
	kindOr
)

// Filter is an immutable RSQL filter expression. Build one with the
// comparison constructors ([Eq], [Ne], [Gt], [Ge], [Lt], [Le], [In], [Out],
// [IsNull], [IsNotNull]) and chain with [Filter.And] / [Filter.Or].
type Filter struct {
	expr string
	kind filterKind
}

// isFilterValueSafe reports whether every char is in [A-Za-z0-9_.:@*+-]
// (the RSQL-safe set; `*` kept bare so wildcards work).
func isFilterValueSafe(s string) bool {
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case c >= 'A' && c <= 'Z', c >= 'a' && c <= 'z', c >= '0' && c <= '9':
		case c == '_', c == '.', c == ':', c == '@', c == '*', c == '+', c == '-':
		default:
			return false
		}
	}
	return true
}

func quoteFilterValue(s string) string {
	var b strings.Builder
	b.Grow(len(s) + 2)
	b.WriteByte('"')
	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '\\':
			b.WriteString(`\\`)
		case '"':
			b.WriteString(`\"`)
		default:
			b.WriteByte(s[i])
		}
	}
	b.WriteByte('"')
	return b.String()
}

// renderFilterValue renders a value for an RSQL expression: numbers and
// booleans bare; strings bare iff non-empty and made only of safe chars,
// otherwise double-quoted with backslash/quote escaping.
func renderFilterValue(v any) string {
	switch x := v.(type) {
	case string:
		return renderFilterString(x)
	case bool:
		return strconv.FormatBool(x)
	case int:
		return strconv.FormatInt(int64(x), 10)
	case int8:
		return strconv.FormatInt(int64(x), 10)
	case int16:
		return strconv.FormatInt(int64(x), 10)
	case int32:
		return strconv.FormatInt(int64(x), 10)
	case int64:
		return strconv.FormatInt(x, 10)
	case uint:
		return strconv.FormatUint(uint64(x), 10)
	case uint8:
		return strconv.FormatUint(uint64(x), 10)
	case uint16:
		return strconv.FormatUint(uint64(x), 10)
	case uint32:
		return strconv.FormatUint(uint64(x), 10)
	case uint64:
		return strconv.FormatUint(x, 10)
	case float32:
		return strconv.FormatFloat(float64(x), 'f', -1, 32)
	case float64:
		return strconv.FormatFloat(x, 'f', -1, 64)
	default:
		return renderFilterString(fmt.Sprint(v))
	}
}

func renderFilterString(s string) string {
	if len(s) > 0 && isFilterValueSafe(s) {
		return s
	}
	return quoteFilterValue(s)
}

func comparison(field, op string, value any) Filter {
	return Filter{expr: field + op + renderFilterValue(value), kind: kindComparison}
}

func list(name, field, op string, values []any) Filter {
	if len(values) == 0 {
		panic("rocksky: " + name + "(\"" + field + "\", ...) needs at least one value")
	}
	rendered := make([]string, len(values))
	for i, v := range values {
		rendered[i] = renderFilterValue(v)
	}
	return Filter{expr: field + op + "(" + strings.Join(rendered, ",") + ")", kind: kindComparison}
}

// Eq is `field==value` — equals; `*` in string values is a wildcard.
func Eq(field string, value any) Filter { return comparison(field, "==", value) }

// Ne is `field!=value` — not equals.
func Ne(field string, value any) Filter { return comparison(field, "!=", value) }

// Gt is `field=gt=value` — greater than.
func Gt(field string, value any) Filter { return comparison(field, "=gt=", value) }

// Ge is `field=ge=value` — greater than or equal.
func Ge(field string, value any) Filter { return comparison(field, "=ge=", value) }

// Lt is `field=lt=value` — less than.
func Lt(field string, value any) Filter { return comparison(field, "=lt=", value) }

// Le is `field=le=value` — less than or equal.
func Le(field string, value any) Filter { return comparison(field, "=le=", value) }

// In is `field=in=(a,b)` — matches any of the values. Panics if values is empty.
func In(field string, values ...any) Filter { return list("In", field, "=in=", values) }

// Out is `field=out=(a,b)` — matches none of the values. Panics if values is empty.
func Out(field string, values ...any) Filter { return list("Out", field, "=out=", values) }

// IsNull is `field==null` — the field is NULL.
func IsNull(field string) Filter { return Filter{expr: field + "==null", kind: kindComparison} }

// IsNotNull is `field!=null` — the field is not NULL.
func IsNotNull(field string) Filter { return Filter{expr: field + "!=null", kind: kindComparison} }

// renderInAnd parenthesizes an OR operand so RSQL precedence is preserved.
func (f Filter) renderInAnd() string {
	if f.kind == kindOr {
		return "(" + f.expr + ")"
	}
	return f.expr
}

// And requires both sides to match (`;`). An OR operand is parenthesized to
// keep RSQL precedence.
func (f Filter) And(other Filter) Filter {
	return Filter{expr: f.renderInAnd() + ";" + other.renderInAnd(), kind: kindAnd}
}

// Or lets either side match (`,`).
func (f Filter) Or(other Filter) Filter {
	return Filter{expr: f.expr + "," + other.expr, kind: kindOr}
}

// Build returns the RSQL expression string to send as the `filter` query param.
func (f Filter) Build() string { return f.expr }

// String returns the RSQL expression string.
func (f Filter) String() string { return f.expr }
